hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/TcpIpJoiner.java

```diff
@@ -483,15 +483,25 @@ public void searchForOtherClusters() {
         SplitBrainJoinMessage request = node.createSplitBrainJoinMessage();
         for (Address address : possibleAddresses) {
             SplitBrainMergeCheckResult result = sendSplitBrainJoinMessageAndCheckResponse(address, request);
-            if (result == SplitBrainMergeCheckResult.LOCAL_NODE_SHOULD_MERGE) {
-                logger.warning(node.getThisAddress() + " is merging [tcp/ip] to " + address);
-                setTargetAddress(address);
-                startClusterMerge(address, request.getMemberListVersion());
-                return;
+             if (result == SplitBrainMergeCheckResult.REMOTE_NODE_SHOULD_MERGE) {
+                String targetAddressStr = address.getHost() + ":" + address.getPort();
+                Address thisAddress = node.address;
+                String thisAddressStr = thisAddress.getHost() + ":" + thisAddress.getPort();
+                if (getMembers().contains(targetAddressStr) && !getMembers().contains(thisAddressStr)) {
```

Comment:
It is necessary to acknowledge that the master member of the other cluster does not know the master member of the local cluster to avoid the scenario that both clusters initiate this split brain merge at the same time. This local check, `!getMembers().contains(thisAddressStr)`, is not sufficient for this.